qa/tasks/ceph_test_case.py

```diff
@@ -326,21 +343,25 @@ def wait_until_true(cls, condition, timeout, check_fn=None, period=5):
         elapsed = 0
         retry_count = 0
         while True:
-            if condition():
-                log.debug("wait_until_true: success in {0}s and {1} retries".format(elapsed, retry_count))
-                return
-            else:
-                if elapsed >= timeout:
-                    if check_fn and check_fn() and retry_count < 5:
-                        elapsed = 0
-                        retry_count += 1
-                        log.debug("wait_until_true: making progress, waiting (timeout={0} retry_count={1})...".format(timeout, retry_count))
-                    else:
-                        raise TestTimeoutError("Timed out after {0}s and {1} retries".format(elapsed, retry_count))
+            if (last_result := condition()):
+                if isinstance(last_result, tuple):
+                    if last_result.result == 0:
+                        log.debug("wait_until_true: success in {0}s and {1} retries".format(elapsed, retry_count))
+                        return
+                elif isinstance(last_result, bool):
+                    log.debug("wait_until_true: success in {0}s and {1} retries".format(elapsed, retry_count))
+                    return
+            if elapsed >= timeout:
+                if check_fn and check_fn() and retry_count < 5:
+                    elapsed = 0
+                    retry_count += 1
+                    log.debug("wait_until_true: making progress, waiting (timeout={0} retry_count={1})...".format(timeout, retry_count))
                 else:
-                    log.debug("wait_until_true: waiting (timeout={0} retry_count={1})...".format(timeout, retry_count))
-                time.sleep(period)
-                elapsed += period
+                    raise TestTimeoutError("Timed out after {0}s and {1} retries: {2}".format(elapsed, retry_count, last_result.context))
```

Comment:
No need to make use of the attribute `.context`. I designed the class `ConditionResult` to behave as a bool or str dendending on the context (`bool` in `if...`, `str` when printed) in order to minimize code changes.

For example, this will break the legacy simple `bool`s. They don't have `.context` property. If we don't want to create 2 paths for `bools` and `ConditionResult`s, we can simply print the bool (not great, but cleaner):
```py
raise TestTimeoutError("Timed out after {0}s and {1} retries: {2}".format(elapsed, retry_count, last_result)) # Works for both bool and ConditionResult
```